File: labs/M02-tokens/solution/cost_estimator.js
/**
 * M02 Lab — Step 2: Cost Estimator (Node.js Solution)
 * =====================================================
 * Predict API costs before making requests.
 */

// Pricing per million tokens (as of 2024)
const PRICING = {
  "claude-haiku": { input: 0.25, output: 1.25 },
  "claude-sonnet": { input: 3.0, output: 15.0 },
  "claude-opus": { input: 15.0, output: 75.0 },
};

/**
 * Estimate the cost of an API call.
 *
 * @param {number} inputTokens - Number of input tokens
 * @param {number} outputTokens - Number of output tokens
 * @param {string} model - Model key from PRICING
 * @returns {{ inputCost: number, outputCost: number, totalCost: number, model: string }}
 */
function estimateCost(inputTokens, outputTokens, model = "claude-sonnet") {
  const prices = PRICING[model];
  if (!prices) {
    throw new Error(`Unknown model: ${model}`);
  }
  const inputCost = (inputTokens / 1_000_000) * prices.input;
  const outputCost = (outputTokens / 1_000_000) * prices.output;
  return {
    inputCost,
    outputCost,
    totalCost: inputCost + outputCost,
    model,
  };
}

// ─── Main ───────────────────────────────────────────────────────────────────

console.log("=== Cost Estimator ===\n");

const scenarios = [
  ["Short query", 100, 200],
  ["Medium query", 1000, 2000],
  ["Batch of 1000", 1000 * 1000, 1000 * 2000],
];

for (const modelKey of ["claude-haiku", "claude-sonnet"]) {
  console.log(`Model: ${modelKey}`);
  for (const [label, inp, out] of scenarios) {
    try {
      const est = estimateCost(inp, out, modelKey);
      console.log(
        `  ${label.padEnd(18)} (${inp} in / ${out} out):   $${est.totalCost.toFixed(6)}`
      );
    } catch (e) {
      console.log(`  ${label.padEnd(18)} [ERROR] ${e.message}`);
    }
  }
  console.log();
}

// Cost comparison
try {
  const haiku = estimateCost(1000, 2000, "claude-haiku");
  const sonnet = estimateCost(1000, 2000, "claude-sonnet");
  const ratio = sonnet.totalCost / haiku.totalCost;
  console.log("Cost comparison:");
  console.log(
    `  Sonnet is ${ratio.toFixed(1)}x more expensive than Haiku for the same workload.`
  );
  const dailyHaiku = haiku.totalCost * 10000;
  const dailySonnet = sonnet.totalCost * 10000;
  console.log(
    `  For 10,000 queries/day, Haiku = $${dailyHaiku.toFixed(2)}/day, Sonnet = $${dailySonnet.toFixed(2)}/day.`
  );
} catch (e) {
  console.log(`  [ERROR] ${e.message}`);
}

File: labs/M02-tokens/starter/cost_estimator.py
"""
M02 Lab — Step 2: Cost Estimator
==================================
Predict API costs before making requests.
"""

# Pricing per million tokens (as of 2024)
PRICING = {
    "claude-haiku": {"input": 0.25, "output": 1.25},
    "claude-sonnet": {"input": 3.00, "output": 15.00},
    "claude-opus": {"input": 15.00, "output": 75.00},
}

def estimate_cost(input_tokens: int, output_tokens: int, model: str = "claude-sonnet") -> dict:
    """
    Estimate the cost of an API call.
    Returns dict with input_cost, output_cost, total_cost, model.
    """
    # TODO:
    # 1. Look up the model's pricing from PRICING dict
    # 2. Calculate input_cost = (input_tokens / 1_000_000) * price_per_million_input
    # 3. Calculate output_cost = (output_tokens / 1_000_000) * price_per_million_output
    # 4. Return {"input_cost": ..., "output_cost": ..., "total_cost": ..., "model": model}
    pass

if __name__ == "__main__":
    print("=== Cost Estimator ===\n")
    scenarios = [
        ("Short query", 100, 200),
        ("Medium query", 1000, 2000),
        ("Batch of 1000", 1000 * 1000, 1000 * 2000),
    ]
    for model_key in ["claude-haiku", "claude-sonnet"]:
        print(f"Model: {model_key}")
        for label, inp, out in scenarios:
            try:
                est = estimate_cost(inp, out, model_key)
                print(f"  {label:<18} ({inp} in / {out} out):   ${est['total_cost']:.6f}")
            except Exception as e:
                print(f"  {label:<18} [ERROR] {e}")
        print()
    # Cost comparison
    try:
        haiku = estimate_cost(1000, 2000, "claude-haiku")
        sonnet = estimate_cost(1000, 2000, "claude-sonnet")
        ratio = sonnet["total_cost"] / haiku["total_cost"]
        print("Cost comparison:")
        print(f"  Sonnet is {ratio:.1f}x more expensive than Haiku for the same workload.")
        daily_haiku = haiku["total_cost"] * 10000
        daily_sonnet = sonnet["total_cost"] * 10000
        print(f"  For 10,000 queries/day, Haiku = ${daily_haiku:.2f}/day, Sonnet = ${daily_sonnet:.2f}/day.")
    except Exception as e:
        print(f"  [ERROR] {e}")
